storage/transfermanager/downloader.go

```diff
@@ -155,46 +173,110 @@ func (d *Downloader) downloadWorker() {
-
-		// Keep track of any error that occurred.
-		if out.Err != nil {
-			d.error(fmt.Errorf("downloading %q from bucket %q: %w", input.Object, input.Bucket, out.Err))
-		}
-
-		// Either execute the callback, or append to results.
-		if d.config.asynchronous {
-			input.Callback(out)
+		out := input.downloadShard(d.client, d.config.perOperationTimeout, d.config.partSize)
+
+		if input.shard == 0 {
+			if out.Err != nil {
+				// Don't queue more shards if the first failed.
+				d.addResult(input, out)
+			} else {
+				numShards := numShards(out.Attrs, input.Range, d.config.partSize)
+
+				if numShards <= 1 {
+					// Download completed with a single shard.
+					d.addResult(input, out)
+				} else {
+					// Queue more shards.
+					outs := d.queueShards(input, out.Attrs.Generation, numShards)
+					// Start a goroutine that gathers shards sent to the output
+					// channel and adds the result once it has received all shards.
+					go d.gatherShards(input, outs, numShards)
+				}
+			}
 		} else {
-			d.addResult(out)
+			// If this isn't the first shard, send to the output channel specific to the object.
+			// This should never block since the channel is buffered to exactly the number of shards.
+			input.shardOutputs <- out
 		}
 	}
 	d.workers.Done()
 }
 
+// queueShards queues all subsequent shards of an object after the first.
+// The results should be forwarded to the returned channel.
+func (d *Downloader) queueShards(in *DownloadObjectInput, gen int64, shards int) <-chan *DownloadOutput {
+	// Create a channel that can be received from to compile the
+	// shard outputs.
+	outs := make(chan *DownloadOutput, shards)
+	in.shardOutputs = outs
+
+	// Create a shared context that we can cancel if a shard fails.
+	in.ctx, in.cancelCtx = context.WithCancelCause(in.ctx)
+
+	// Add generation in case the object changes between calls.
+	in.Generation = &gen
+
+	// Queue remaining shards.
+	// We could optimize for zero-initialized DownloadBuffers if we queue the last
+	// shard first (so the buffer grows to the appropriate length),
+	// but that might affect file performance (if implementation
+	// prefers ordered writes).
+	for i := 1; i < shards; i++ {
+		newShard := in // this is fine, since the input should only differ in the shard num
+		newShard.shard = i
+		d.addInput(newShard)
+	}
+
+	return outs
+}
+
+var errCancelAllShards = errors.New("cancelled because another shard failed")
+
+// gatherShards receives from the given channel exactly (shards-1) times (since
+// the first shard should already be complete).
+// It will add the result to the Downloader once it has received all shards.
+// gatherShards cancels remaining shards if any shard errored.
+func (d *Downloader) gatherShards(in *DownloadObjectInput, outs <-chan *DownloadOutput, shards int) {
+	errs := []error{}
+	var shardOut *DownloadOutput
+	for i := 1; i < shards; i++ {
+		// Add monitoring here? This could hang if any individual piece does.
```

Comment:
By monitoring do you mean metrics or some kind of go routine to monitor?

Reply:
A go routine, but it might not be necessary as everything should cancel when the ctx does (so timeouts would stop this from hanging)

Reply:
Okay, I think as long as we have a test for a hanging part then we can remove the comment.

Reply:
I'm not sure if there is an easy way to simulate a hanging part, besides that would just be testing that a read request gets cancelled when its context is, which feels out of scope for TM. 

Other than that, if we add some logic here then we could test - some options:
1. extend the perOpTimeout and stop listening for pieces when done (start a timer with per_op_timeout * num_shards) 
2. have a default timer for collecting all pieces (probably not this)
3. check for ctx cancel directly in here and stop when cancelled, but I don't like that because then it would miss receiving any other errors in shards that were just completing...
4. instead of 3 with immediate cancellation, maybe a timer for a couple seconds could start once ctx cancel is received, and we only stop listening once that timer is up